lucetc/src/compiler/cpu_features.rs

```diff
@@ -0,0 +1,159 @@
+use crate::error::{LucetcError, LucetcErrorKind};
+use cranelift_codegen::{isa, settings::Configurable};
+use failure::{format_err, ResultExt};
+use target_lexicon::Triple;
+
+/// x86 CPU families used as shorthand for different CPU feature configurations.
+///
+/// Matches the definitions from `cranelift-codegen`'s x86 settings definition.
+#[derive(Debug, Clone, Copy)]
+pub enum TargetCpu {
+    Baseline,
+    Nehalem,
+    Haswell,
+    Broadwell,
+    Skylake,
+    Cannonlake,
+    Icelake,
+    Znver1,
+}
+
+/// A manual specification of the CPU features to use during codegen.
+#[derive(Debug, Clone, Copy)]
+pub struct SpecificFeatures {
+    pub has_sse3: bool,
+    pub has_ssse3: bool,
+    pub has_sse41: bool,
+    pub has_sse42: bool,
+    pub has_popcnt: bool,
+    pub has_avx: bool,
+    pub has_bmi1: bool,
+    pub has_bmi2: bool,
+    pub has_lzcnt: bool,
+}
+
+impl SpecificFeatures {
+    /// Return a `CpuFeatures` with all optional features disabled.
+    pub fn all_disabled() -> Self {
+        Self {
+            has_sse3: false,
+            has_ssse3: false,
+            has_sse41: false,
+            has_sse42: false,
+            has_popcnt: false,
+            has_avx: false,
+            has_bmi1: false,
+            has_bmi2: false,
+            has_lzcnt: false,
+        }
+    }
+}
+
+impl From<TargetCpu> for SpecificFeatures {
+    fn from(cpu: TargetCpu) -> Self {
+        use TargetCpu::*;
+        match cpu {
+            Baseline => Self::all_disabled(),
+            Nehalem => Self {
+                has_sse3: true,
+                has_ssse3: true,
+                has_sse41: true,
+                has_sse42: true,
+                has_popcnt: true,
+                ..Baseline.into()
+            },
+            Haswell => Self {
+                // Note: this is not part of the Cranelift profile for Haswell, which only uses
+                // CPUID detection to enable AVX. If we want to bypass CPUID when compiling, we need
+                // to set it manually, and Haswell is the first of the CPUs with profiles to have
+                // AVX.
+                has_avx: true,
```

Comment:
I'm so shiny and new that I don't expect to contribute meaningfully to PR review.  That said, I do like to treat PR review as an opportunity for learning.  When I look at the CPU settings in crane lift, I don't see that it `has-avx`. 

https://github.com/CraneStation/cranelift/blob/c053a0bbe1fec8c23ad6f15148e0b26efe194f7c/cranelift-codegen/meta/src/isa/x86/settings.rs#L81-L83

 I wonder how I would learn that haswell has the AVX instruction, as you've indicated here, aside from glancing at a Wikipedia page (like I did over at: https://en.wikipedia.org/wiki/Advanced_Vector_Extensions).

Reply:
It still is detecting and using AVX further up in that file: https://github.com/CraneStation/cranelift/blob/c053a0bbe1fec8c23ad6f15148e0b26efe194f7c/cranelift-codegen/meta/src/isa/x86/settings.rs#L12

It's strange, though, that it doesn't include it in the CPU family profiles.

Also, it took a little bit to find, but an x86_64 gcc's manpage gives a pretty good rundown on what these aliases entail: https://embarc.org/man-pages/gcc/x86-Options.html

Reply:
Thanks for this guidance!  ❤️   Thanks, too, for validating the strangeness that AVX isn't included in the CPU family profiles.  Please let me know if it's alright to ask these sorts of questions on PRs, or if slack is a better place for these kind of "slightly out of scope" conversations.